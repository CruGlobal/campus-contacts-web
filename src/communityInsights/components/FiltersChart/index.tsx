import React, { useState } from 'react';
import { useQuery } from 'react-apollo-hooks';
import moment from 'moment';
import _ from 'lodash';

import BarChart from '../BarChart';
import NullState from '../NullState';

interface Props {
    query: any;
    variables?: any;
    mapData: (data: any) => any;
    label: string;
    currentDate?: Date;
    nullContent: string;
}

const FiltersChart = ({
    query,
    variables,
    mapData,
    label,
    currentDate,
    nullContent,
}: Props) => {
    const [index, setIndex] = useState(0);
    const [endDate, setEndDate] = useState(moment(currentDate));
    const [type, setType] = useState('month');

    const { data, loading } = useQuery(query, {
        variables: {
            ...variables,
            period: type === 'month' ? 'P1M' : 'P1Y',
            endDate,
        },
    });

    const checkForNullContent = (data: any) => {
        const nullCheck = data.community.report.daysReport.filter(
            (report: any) => {
                return (
                    report.personalStepsAddedCount !== 0 &&
                    report.othersStepsAddedCount !== 0 &&
                    report.interactionsCount !== 0 &&
                    report.personalStepsCompletedCount !== 0 &&
                    report.othersStepsCompletedCount !== 0
                );
            },
        );

        return nullCheck.length === 0;
    };

    if (loading) {
        return <NullState content={nullContent} />;
    }

    const updateDates = (type: string, index: number) => {
        setType(type);
        setIndex(index);
        setEndDate(
            moment(currentDate).subtract(
                index,
                type === 'month' ? 'months' : 'years',
            ),
        );
    };

    const mapDataDaily = (data: any) => {
        const graphData = [];
        const end = moment(endDate);
        const start = moment(endDate).subtract(1, 'months');

        for (let m = moment(start); m.isBefore(end); m.add(1, 'days')) {
            const exists = _.find(data, { date: m.format('YYYY-MM-DD') });
            graphData.push({
                ...exists,
                date: m.format('DD'),
            });
        }
        return graphData;
    };

    const mapDataMonthly = (data: any) => {
        const graphData = [];
        const end = moment(endDate).endOf('month');
        const start = moment(endDate)
            .subtract(11, 'months')
            .startOf('month');

        for (let m = start; m.isBefore(end); m.add(1, 'months')) {
            const endOfMonth = moment(m).endOf('month');
            const exists = _.filter(data, e =>
                moment(e.date).isBetween(m, endOfMonth),
            );
            const reduced = _.reduce(exists, (result, current) => {
                return {
                    ...result,
                    total: result.total + current.total,
                    stages: [...result.stages, ...current.stages],
                };
            });
            graphData.push({
                ...reduced,
                date: m.format('MM'),
            });
        }
        return graphData;
    };

    const mapDataToType = (data: any) => {
        if (type === 'month') {
            return mapDataDaily(data);
        }
        return mapDataMonthly(data);
    };

    const mappedData = mapData(data);
    const graphData = mapDataToType(mappedData);
    const average = Math.floor(_.sumBy(graphData, 'total') / graphData.length);

    return (
        <BarChart
            nullContent={nullContent}
            nullCheck={checkForNullContent(data)}
            data={graphData}
            keys={['total']}
            indexBy={'date'}
            average={average}
            tooltipBreakdown={true}
            datesFilter={true}
            index={index}
            filterType={type}
            onFilterChanged={(type, index) => updateDates(type, index)}
            legendLabel={label}
        />
    );
};

export default FiltersChart;
