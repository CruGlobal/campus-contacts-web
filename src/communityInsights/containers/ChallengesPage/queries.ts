import gql from 'graphql-tag';

const GET_IMPACT_CHALLENGES = gql`
    query impactReportStepsCount($id: ID!) {
        community(id: $id) {
            impactReport {
                stepsCount
            }
        }
    }
`;

const GET_CHALLENGES = gql`
    query communityChallenges(
        $id: ID!
        $first: Int
        $last: Int
        $after: String
        $before: String
        $sortBy: [CommunityChallengesSortEnum!]
    ) {
        community(id: $id) {
            communityChallenges(
                first: $first
                last: $last
                after: $after
                before: $before
                sortBy: $sortBy
            ) {
                nodes {
                    id
                    title
                    acceptedCount
                    completedCount
                    createdAt
                    endDate
                }
                pageInfo {
                    hasNextPage
                    startCursor
                    endCursor
                    hasPreviousPage
                }
            }
        }
    }
`;

export { GET_CHALLENGES, GET_IMPACT_CHALLENGES };
