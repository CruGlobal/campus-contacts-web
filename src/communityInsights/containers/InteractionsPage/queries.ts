import gql from 'graphql-tag';

const INTERACTIONS_TOTAL_REPORT = gql`
    query impactReportInteractionsCount($id: ID!) {
        community(id: $id) {
            impactReport {
                interactionsCount
                interactionsReceiversCount
            }
        }
    }
`;

const INTERACTIONS_TOTAL_COMPLETED_REPORT = gql`
    query communityReportInteractions(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                interactions {
                    nodes {
                        interactionCount
                        interactionType {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const INTERACTIONS_COMPLETED_REPORT = gql`
    query communityReportDaysInteractions(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                daysReport {
                    nodes {
                        date
                        interactionsCount
                        interactionResults {
                            nodes {
                                count
                                interactionType {
                                    name
                                }
                            }
                        }
                    }
                }
            }
        }
    }
`;

export {
    INTERACTIONS_TOTAL_REPORT,
    INTERACTIONS_TOTAL_COMPLETED_REPORT,
    INTERACTIONS_COMPLETED_REPORT,
};
