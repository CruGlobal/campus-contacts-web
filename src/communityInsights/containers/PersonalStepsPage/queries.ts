import gql from 'graphql-tag';

const GET_IMPACT_REPORT_MOVED = gql`
    query impactReportStageProgressionCount($id: ID!) {
        community(id: $id) {
            impactReport {
                stageProgressionCount
            }
        }
    }
`;

const GET_IMPACT_REPORT_STEPS_TAKEN = gql`
    query impactReportPersonalStepsCompletedCount($id: ID!) {
        community(id: $id) {
            impactReport {
                personalStepsCompletedCount
            }
        }
    }
`;

const GET_STAGES_REPORT_MEMBER_COUNT = gql`
    query communityReportStagesPersonalMemberCount(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        memberCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_STAGES_REPORT_STEPS_ADDED = gql`
    query communityReportStagesPersonalStepsAdded(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        personalStepsAddedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_STEPS_COMPLETED_REPORT = gql`
    query communityReportDaysPersonalSteps(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                daysReport {
                    nodes {
                        date
                        personalStepsCompletedCount
                        stageResults {
                            nodes {
                                personalStepsCompletedCount
                                stage {
                                    name
                                }
                            }
                        }
                    }
                }
            }
        }
    }
`;

const GET_TOTAL_STEPS_COMPLETED_SUMMARY = gql`
    query communityReportStagesPersonalStepsCompleted(
        $id: ID!
        $period: String!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        personalStepsCompletedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

export {
    GET_IMPACT_REPORT_MOVED,
    GET_IMPACT_REPORT_STEPS_TAKEN,
    GET_STAGES_REPORT_MEMBER_COUNT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT_STEPS_ADDED,
    GET_TOTAL_STEPS_COMPLETED_SUMMARY,
};
