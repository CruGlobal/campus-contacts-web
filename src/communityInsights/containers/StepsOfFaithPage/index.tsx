import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import AppContext from '../../../appContext';
import StagesSummary from '../../components/StagesSummary';
import StepsChart from '../../components/StepsChart';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    GET_IMPACT_REPORT_TAKEN,
    GET_TOTAL_STEPS_COMPLETED_REPORT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT,
    GET_IMPACT_REPORT_REACHED,
    GET_STAGES_PEOPLE_REPORT,
} from './queries';

const StepsOfFaithPage = () => {
    const { t } = useTranslation('insights');
    const { orgId }: any = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_REPORT_TAKEN}
                text={report =>
                    t('stepsOfFaith.taken', {
                        count:
                            report.community.impactReport
                                .othersStepsCompletedCount,
                        people:
                            report.community.impactReport
                                .othersStepsReceiversCompletedCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card title={t('stepsOfFaith.totalCompleted')}>
                <StagesSummary
                    query={GET_TOTAL_STEPS_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={data =>
                        data.community.report.stagesReport.map(
                            (entry: any) => ({
                                stage: entry.stage.name,
                                icon: entry.stage.name
                                    .toLowerCase()
                                    .replace(' ', '-'),
                                count: entry.othersStepsCompletedCount,
                            }),
                        )
                    }
                />
            </Card>
            <Card
                title={t('stepsOfFaith.completed')}
                subtitle={t('stepsOfFaith.completedSubtitle')}
            >
                <FiltersChart
                    query={GET_STEPS_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={data =>
                        data.community.report.daysReport.map((row: any) => ({
                            ['total']: row.othersStepsCompletedCount,
                            ['stages']: row.stageResults.map((stage: any) => ({
                                name: stage.othersStepsCompletedCount,
                                count: stage.stage.name,
                            })),
                            ['date']: row.date,
                        }))
                    }
                    currentDate={moment().toDate()}
                    label={t('stepsOfFaith.legend')}
                />
            </Card>
            <Card
                title={t('stepsOfFaith.added')}
                subtitle={t('stepsOfFaith.addedSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_REPORT}
                    mapData={data =>
                        data.community.report.stagesReport.map((row: any) => ({
                            [t(
                                'stepsOfFaith.legendLabel',
                            )]: row.othersStepsAddedCount,
                            [t('stage')]: row.stage.name.toUpperCase(),
                        }))
                    }
                    label={t('stepsOfFaith.legendLabel')}
                    index={t('stage')}
                    variables={{
                        period: 'P10Y',
                        id: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>
            <ImpactInfo
                query={GET_IMPACT_REPORT_REACHED}
                text={report =>
                    t('stepsOfFaith.reached', {
                        count:
                            report.community.impactReport
                                .membersStageProgressionCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card
                title={t('stepsOfFaith.people')}
                subtitle={t('stepsOfFaith.peopleSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_PEOPLE_REPORT}
                    mapData={data =>
                        data.community.report.stagesReport.map((row: any) => ({
                            [t('stepsOfFaith.peopleLabel')]: row.contactCount,
                            [t('stage')]: row.stage.name.toUpperCase(),
                        }))
                    }
                    label={t('stepsOfFaith.peopleLabel')}
                    index={t('stage')}
                    variables={{
                        period: 'P10Y',
                        id: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>
        </div>
    );
};

export default StepsOfFaithPage;
