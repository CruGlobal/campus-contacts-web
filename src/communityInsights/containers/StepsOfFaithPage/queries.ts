import gql from 'graphql-tag';

const GET_IMPACT_REPORT_TAKEN = gql`
    query impactReportOtherStepsCompleted($id: ID!) {
        community(id: $id) {
            impactReport {
                othersStepsCompletedCount
                othersStepsReceiversCompletedCount
            }
        }
    }
`;

const GET_TOTAL_STEPS_COMPLETED_REPORT = gql`
    query communityReportStagesOthersStepsCompleted(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        othersStepsCompletedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_STEPS_COMPLETED_REPORT = gql`
    query communityReportDaysOthersSteps(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                daysReport {
                    nodes {
                        date
                        othersStepsCompletedCount
                        stageResults {
                            nodes {
                                othersStepsCompletedCount
                                stage {
                                    name
                                }
                            }
                        }
                    }
                }
            }
        }
    }
`;

const GET_STAGES_REPORT = gql`
    query communityReportStagesOthersStepsAdded(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        othersStepsAddedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_IMPACT_REPORT_REACHED = gql`
    query impactReportMembersStageProgressionCount($id: ID!) {
        community(id: $id) {
            impactReport {
                membersStageProgressionCount
            }
        }
    }
`;

const GET_STAGES_PEOPLE_REPORT = gql`
    query communityReportStagesOthersContactCount(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        contactCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

export {
    GET_IMPACT_REPORT_TAKEN,
    GET_IMPACT_REPORT_REACHED,
    GET_TOTAL_STEPS_COMPLETED_REPORT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT,
    GET_STAGES_PEOPLE_REPORT,
};
