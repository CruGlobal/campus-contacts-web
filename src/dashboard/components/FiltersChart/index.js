import React, { useState } from 'react';
import { useQuery } from 'react-apollo-hooks';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import _ from 'lodash';

import BarChart from '../BarChart';

const FiltersChart = ({
    query,
    variables,
    mapData,
    countAverage,
    label,
    currentDate,
}) => {
    const [index, setIndex] = useState(0);
    const [endDate, setEndDate] = useState(moment(currentDate));
    const [type, setType] = useState('month');

    const { data, loading } = useQuery(query, {
        variables: {
            ...variables,
            period: type === 'month' ? 'P1M' : 'P1Y',
            endDate,
        },
    });
    const { t } = useTranslation('insights');

    if (loading) {
        return <div>{t('loading')}</div>;
    }

    const updateDates = (type, index) => {
        setType(type);
        setIndex(index);
        setEndDate(
            moment(currentDate).subtract(
                index,
                type === 'month' ? 'months' : 'years',
            ),
        );
    };

    const mapDataDaily = data => {
        const graphData = [];
        const end = moment(endDate);
        const start = moment(endDate).subtract(1, 'months');

        for (let m = moment(start); m.isBefore(end); m.add(1, 'days')) {
            const exists = _.find(data, { date: m.format('YYYY-MM-DD') });
            graphData.push({
                ...exists,
                date: m.format('DD'),
            });
        }
        return graphData;
    };

    const mapDataMonthly = data => {
        const graphData = [];
        const end = moment(endDate).endOf('month');
        const start = moment(endDate)
            .subtract(11, 'months')
            .startOf('month');

        for (let m = start; m.isBefore(end); m.add(1, 'months')) {
            const endOfMonth = moment(m).endOf('month');
            const exists = _.filter(data, e =>
                moment(e.date).isBetween(m, endOfMonth),
            );
            const reduced = _.reduce(exists, (result, current) => {
                return {
                    ...result,
                    total: result.total + current.total,
                    stages: [...result.stages, ...current.stages],
                };
            });
            graphData.push({
                ...reduced,
                date: m.format('MM'),
            });
        }
        return graphData;
    };

    const mapDataToType = data => {
        if (type === 'month') {
            return mapDataDaily(data);
        }
        return mapDataMonthly(data);
    };

    const mappedData = mapData(data);
    const graphData = mapDataToType(mappedData);
    const average = Math.floor(_.sumBy(graphData, 'total') / graphData.length);

    return (
        <BarChart
            data={graphData}
            keys={['total']}
            indexBy={'date'}
            average={average}
            tooltipBreakdown={true}
            datesFilter={true}
            index={index}
            filterType={type}
            onFilterChanged={(type, index) => updateDates(type, index)}
            legendLabel={label}
        />
    );
};

export default FiltersChart;
