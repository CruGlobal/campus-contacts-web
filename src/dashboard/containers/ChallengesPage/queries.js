import gql from 'graphql-tag';

const GET_IMPACT_CHALLENGES = gql`
    query communityReport {
        communityReport {
            impactReport {
                interactionsCount
                interactionsReceiversCount
                challengesCount
            }
        }
    }
`;

const GET_CHALLENGES = gql`
    query globalCommunityChallenges($id: ID!, $first: Int, $after: String) {
        community(id: $id) {
            communityChallenges(first: $first, after: $after) {
                nodes {
                    id
                    title
                    acceptedCount
                    completedCount
                    createdAt
                    endDate
                }
                pageInfo {
                    hasNextPage
                    startCursor
                    endCursor
                    hasPreviousPage
                }
            }
        }
    }
`;

export { GET_CHALLENGES, GET_IMPACT_CHALLENGES };
