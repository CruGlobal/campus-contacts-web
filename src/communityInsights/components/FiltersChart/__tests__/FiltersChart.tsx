import React from 'react';
import gql from 'graphql-tag';

import { renderWithContext } from '../../../../testUtils';
import { GET_STEPS_COMPLETED_REPORT } from '../../../containers/StepsOfFaithPage/queries';
import FiltersChart from '..';

describe('<FiltersChart />', () => {
    it('should render properly', async () => {
        const { snapshot, unmount } = renderWithContext(
            <FiltersChart
                nullContent={'stepsOfFaithCompleted'}
                query={GET_STEPS_COMPLETED_REPORT}
                mapData={() => {}}
                label={'test label'}
            />,
        );
        snapshot();
        unmount();
    });

    // TODO: Add a test with mocking data when GraphQL schema contains above query
});
