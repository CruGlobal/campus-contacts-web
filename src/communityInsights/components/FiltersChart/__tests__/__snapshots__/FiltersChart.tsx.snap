// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersChart /> should render properly 1`] = `
.emotion-0 {
  width: 100%;
  height: 400px;
  background-color: #ffffff;
  -webkit-animation: fade 2s infinite linear both;
  animation: fade 2s infinite linear both;
  border-radius: 8px;
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;
